fedimint-core/src/task/jit.rs

```diff
@@ -87,89 +86,42 @@ where
-                    _ = cancel_rx.recv() => {
-                        trace!(target: LOG_TASK, r#type = %type_name, "JitTry value future canceled");
-                    },
-                    val = f() => {
-                        match val_tx.send(val) {
-                            Ok(_) => {
-                                trace!(target: LOG_TASK, r#type = %type_name, "JitTry value ready");
-                            },
-                            Err(_) => {
-                                trace!(target: LOG_TASK,  r#type = %type_name, "JitTry value ready, but ignored");
-                            },
-                        };
-                    },
-                }
-            },
+            async move { f().await },
         );
 
         Self {
-            _cancel_tx: cancel_tx,
-            val_rx: Arc::new(Some(val_rx).into()),
-            val: sync::OnceCell::new().into(),
+            inner: JitInner {
+                handle: handle.into(),
+                val: sync::OnceCell::new(),
+            }
+            .into(),
         }
     }
 
     /// Get the reference to the value, potentially blocking for the
     /// initialization future to complete
-    pub async fn get_try(&self) -> std::result::Result<&T, OneTimeError<E>> {
-        /// Temporarily taken data out of `arc`, that will be put back on drop,
-        /// unless canceled
-        ///
-        /// Used to achieve cancelation safety.
-        struct PutBack<'a, T> {
-            val: Option<T>,
-            arc: &'a Arc<std::sync::Mutex<Option<T>>>,
-        }
-
-        impl<'a, T> PutBack<'a, T> {
-            fn take(arc: &'a Arc<std::sync::Mutex<Option<T>>>) -> Self {
-                let val = arc.lock().expect("lock failed").take();
-                Self { val, arc }
-            }
-
-            fn cancel(mut self) {
-                self.val = None;
-            }
-
-            fn get_mut(&mut self) -> Option<&mut T> {
-                self.val.as_mut()
-            }
-        }
-        impl<'a, T> Drop for PutBack<'a, T> {
-            fn drop(&mut self) {
-                let mut lock = self.arc.lock().expect("lock failed");
-                let take = self.val.take();
-
-                *lock = take;
-            }
-        }
-        self.val
-            .get_or_try_init(|| async {
-                let mut recv = PutBack::take(&self.val_rx);
-
-                let val_res = {
-                    let Some(recv) = recv.get_mut() else {
-                        return Err(OneTimeError(None));
-                    };
-                    recv.await.unwrap_or_else(|_| {
-                        panic!("Jit value {} panicked", std::any::type_name::<T>())
+    pub async fn get_try(&self) -> Result<&T, OneTimeError<E>> {
+        let mut init_error = None;
+        let value = self
+            .inner
+            .val
+            .get_or_init(|| async {
+                let handle: &mut _ = &mut *self.inner.handle.lock().await;
```

Comment:
Actually ... no. It moves the handle out here, and there's a `await` 2 lines later where the handle could be dropped.

AFAIK `Drop` is the only reliable tool against cancellation safety.

Reply:
are you sure, it doesn't move out it

handle: &mut JoinHandle

and tokio says `&mut JoinHandle` is cancellation safe 
https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#cancel-safety

Reply:
if &mut JoinHandle poll becomes Ready, we don't have any more .await points, it will always preserve the error.

Reply:
Ah. Nevermind. I got confused, seen things that weren't there.

Reply:
the old drop guard code is round-about way to get `&mut Fut`

Reply:
Now that I look at it, I don't understand how is it possible to await the result of the future on `&mut JoinHandle`, without consuming it. What happens if you call it again?  :D

Reply:
> Now that I look at it, I don't understand how is it possible to await the result of the future on &mut JoinHandle, without consuming it. What happens if you call it again? :D

`Future::poll` only takes `Pin<&mut Self>`.

from https://doc.rust-lang.org/stable/std/future/trait.Future.html

> Once a future has completed (returned Ready from poll), calling its poll method again may panic, block forever, or cause other kinds of problems; the Future trait places no requirements on the effects of such a call. However, as the poll method is not marked unsafe, Rust’s usual rules apply: calls must never cause undefined behavior (memory corruption, incorrect use of unsafe functions, or the like), regardless of the future’s state.

In practice, good libraries always have FusedFutures, i.e. they return `Pending` after they become `Ready` once.